### PR TITLE
Improve error handling if reporting service implementation is missing [FINERACT-1173]

### PR DESCRIPTION
**Pull request by vorburger**
see FINERACT-1173 (for FINERACT-1127)

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/RunreportsApiResource.java [hunk 1]
```diff
@@ -124,12 +125,13 @@ public Response runReport(@PathParam("reportName") @Parameter(description = "rep
             parameterTypeValue = "report";
             String reportType = this.readExtraDataAndReportingService.getReportType(reportName, isSelfServiceUserReport);
             ReportingProcessService reportingProcessService = this.reportingProcessServiceProvider.findReportingProcessService(reportType);
-            if (reportingProcessService != null) {
-                return reportingProcessService.processRequest(reportName, queryParams);
+            if (reportingProcessService == null) {
+                throw new PlatformServiceUnavailableException("err.msg.report.service.implementation.missing",
+                        ReportingProcessServiceProvider.SERVICE_MISSING + reportType, reportType);
```
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/RunreportsApiResource.java]: Unfortunately based on my read of the code, I don't think this is quite right:

After this change, we are bailing out if we can't find a reporting process service for the report type. 
Previously, if we couldn't find one, we would continue to execute the "hardcoded" reporting process service in the same function below. 

I think continuing execution is the expected behaviour: for any report that is not of type Pentaho or SMS, the report_type is set to Table. But there is no ReportingProcessService with type Table, so therefore the logic in this function below is used to simply prepare the CSV / PDF file. 

In other words: if we check this in, as far as I can see it would break any report that doesn't have a specific report handler configured. 

Now having said that, a good idea would be to create a "Table" type report handler and move the code from this function to that report handler. And after that we could assume that there's a handler for each type of report, and bail out if we can't find one.

**Review by ptuomola — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/RunreportsApiResource.java (hunk 1 above)
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/RunreportsApiResource.java]: Oh... wow, very well spotted. I'll have a go at the re-factoring you're suggesting, that would make a lot of sense, yes.

But before I get to that, I'm struggling with more fundamental classloading issues... urgh!! More soon on JIRA, hopefully.

Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/RunreportsApiResource.java]: @ptuomola I'm finally (7 weeks later, phew!) picking this up, and I'm struggling to understand how the `reportType: SMS` is intended to work... @vidakovic do you understand that? Can those be run on the `/runreports` API? I think what I'll do is just write a `DefaultReportingProcessService` with the current code to which this will delegate to for both `Table` and `SMS`.. I'll therefore make the `type` on `org.apache.fineract.infrastructure.report.annotation.ReportService` a multivalue. Does that sounds reasonable to you guys?

Comment (vidakovic) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/RunreportsApiResource.java]: Also not sure how this SMS thing is supposed to work... will have a look at this later today.

Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/RunreportsApiResource.java]: Actually I guess that one would then more appropriately be our `DatatableReportingProcessService` (in `infrastructure.dataqueries.service`) instead of a `DefaultReportingProcessService` in  `infrastructure.report.service.ReportingProcessService`.... watch for an update on this PR! :smile_cat:

**Comment by vorburger** [comment]
@ptuomola re-review this, better now? @vidakovic do you also want to co-pair-review this? Let's give Petri a chance to look it over, so don't merge, just LGTM; this one is a big enough change to warrant 4 extra eyes.. this was supposed to be "just adding a check" but turned into a fairly big refactoring! :smile:

**Review by vidakovic — APPROVED** [review]
LGTM

**Review by ptuomola — APPROVED** [review]
LGTM... sorry for the delay, moving across the world with family in the times of a pandemic is not easy!
